tensorboard/plugins/beholder/video_writing.py

```diff
@@ -12,4 +12,189 @@
+
+  def finish(self):
+    if self.output:
+      self.output.close()
+    self.output = None
+    self.frame_shape = None
+    # Reconsider failed outputs when video is manually restarted.
+    self.output_index = 0
+
+
+class VideoOutput(object):
+  """Base class for video outputs supported by VideoWriter."""
+
+  __metaclass__ = abc.ABCMeta
+
+  # Would add @abc.abstractmethod in python 3.3+
+  @classmethod
+  def available(cls):
+    raise NotImplementedError()
+
+  @classmethod
+  def name(cls):
+    return cls.__name__
+
+  @abc.abstractmethod
+  def emit_frame(self, np_array):
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def close(self):
+    raise NotImplementedError()
+
+
+class PNGVideoOutput(VideoOutput):
+  """Video output implemented by writing individual PNGs to disk."""
+
+  @classmethod
+  def available(cls):
+    return True
+
+  def __init__(self, directory, frame_shape):
+    del frame_shape  # unused
+    self.directory = directory + '/video-frames-{}'.format(time.time())
+    self.frame_num = 0
+    tf.gfile.MakeDirs(self.directory)
+
+  def emit_frame(self, np_array):
+    filename = self.directory + '/{:05}.png'.format(self.frame_num)
+    im_util.write_image(np_array.astype(np.uint8), filename)
+    self.frame_num += 1
+
+  def close(self):
+    pass
+
+
+class FFmpegVideoOutput(VideoOutput):
+  """Video output implemented by streaming to FFmpeg with .mp4 output."""
+
+  @classmethod
+  def available(cls):
+    # Silently check if ffmpeg is available.
+    try:
+      with open(os.devnull, 'wb') as devnull:
+        subprocess.check_call(
+            ['ffmpeg', '-version'], stdout=devnull, stderr=devnull)
+      return True
+    except (OSError, subprocess.CalledProcessError):
+      return False
+
+  def __init__(self, directory, frame_shape):
+    self.filename = directory + '/video-{}.webm'.format(time.time())
+    if len(frame_shape) != 3:
+      raise ValueError(
+          'Expected rank-3 array for frame, got %s' % str(frame_shape))
+    # Set input pixel format based on channel count.
+    if frame_shape[2] == 1:
+      pix_fmt = 'gray'
+    elif frame_shape[2] == 3:
+      pix_fmt = 'rgb24'
+    else:
+      raise ValueError('Unsupported channel count %d' % frame_shape[2])
+
+    command = [
+        'ffmpeg',
+        '-y',  # Overwite output
+        # Input options - raw video file format and codec.
+        '-f', 'rawvideo',
+        '-vcodec', 'rawvideo',
+        '-s', '%dx%d' % (frame_shape[1], frame_shape[0]),  # Width x height.
+        '-pix_fmt', pix_fmt,
+        '-r', '15',  # Frame rate: arbitrarily use 15 frames per second.
+        '-i', '-',  # Use stdin.
+        '-an',  # No audio.
+        # Output options - use lossless VP9 codec inside .webm.
+        '-vcodec', 'libvpx-vp9',
+        '-lossless', '1',
+        # Using YUV is most compatible, though conversion from RGB skews colors.
+        '-pix_fmt', 'yuv420p',
+        self.filename
+    ]
+    PIPE = subprocess.PIPE
```

Comment:
nit question: Just curious, why `subprocess.PIPE` over say `sys.stdout` and `sys.stderr`? Does `PIPE` enable the output of this TensorBoard process to be piped into other processes?

Reply:
PIPE enables the stdout and stderr to be retrieved in the communicate() call later on.  If we used sys.stdout and sys.stderr then anything output by FFmpeg would instead end up in the actual calling code's stdout and stderr.  It seemed better to capture that output ourselves so we can give it more context (i.e. labeling it as coming from FFmpeg) and route it through the logging system.

Strictly speaking we're only using the stderr result so stdout could alternatively be assigned to a devnull file like I do in available(), but we wouldn't be able to use a context manager for the devnull file since it needs to stay open until the VideoOutput itself is closed, which just makes it another resource to manage, so using PIPE seemed easier even though we don't recover that data later.  Note that I don't actually think FFmpeg typically writes anything at all to stdout.

Reply:
Got it!